src/flow/src/compute/render/reduce.rs

```
@@ -0,0 +1,957 @@
+                                arr
+                            })
+                            .collect(),
+                    )
+                }
+            }
+        }
+    }
+}
+
+/// All arrange(aka state) used in reduce operator
+pub struct ReduceArrange {
+    /// The output arrange of reduce operator
+    output_arrange: ArrangeHandler,
+    /// The distinct input arrangement for accumulable reduce plan
+    /// only used when accumulable reduce plan has distinct aggregation
+    distinct_input: Option<Vec<ArrangeHandler>>,
+}
+
+/// split a row into key and val by evaluate the key and val plan
+fn split_row_to_key_val(
+    row: Row,
+    sys_time: repr::Timestamp,
+    diff: repr::Diff,
+    key_val_plan: &KeyValPlan,
+    row_buf: &mut Row,
+) -> Result<Option<KeyValDiffRow>, EvalError> {
+    if let Some(key) = key_val_plan
+        .key_plan
+        .evaluate_into(&mut row.inner.clone(), row_buf)?
+    {
+        // val_plan is not supported to carry any filter predicate,
+        let val = key_val_plan
+            .val_plan
+            .evaluate_into(&mut row.inner.clone(), row_buf)?
+            .with_context(|| InternalSnafu {
+                reason: "val_plan should not contain any filter predicate",
+            })?;
+        Ok(Some(((key, val), sys_time, diff)))
+    } else {
+        Ok(None)
+    }
+}
+
+/// reduce subgraph, reduce the input data into a single row
+/// output is concat from key and val
+fn reduce_subgraph(
+    ReduceArrange {
+        output_arrange: arrange,
+        distinct_input,
+    }: &ReduceArrange,
+    data: impl IntoIterator<Item = DiffRow>,
+    key_val_plan: &KeyValPlan,
+    reduce_plan: &ReducePlan,
+    SubgraphArg {
+        now,
+        err_collector,
+        scheduler,
+        send,
+    }: SubgraphArg,
+) {
+    let mut row_buf = Row::empty();
+    let key_val = data.into_iter().filter_map(|(row, sys_time, diff)| {
+        // error is collected and then ignored
```

Could ignoring the errors make it difficult to address certain runtime issues?

---

> Could ignoring the errors make it difficult to address certain runtime issues?

All `EvalError` are collected in `err_collector` and should be send to be logging, but the impl of logging it is still WIP